File: examples/matmul_example.py
import numpy as np
import time

def bala(A, B):
    n = len(A)
    C = [[0.0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C

def main():
    N = 2048  
    print(f"Starting Python matrix multiplication of size {N}x{N}...")
    A = np.ones((N, N), dtype=np.float32).tolist()
    B = np.ones((N, N), dtype=np.float32).tolist()

    start = time.time()
    C = bala(A, B)
    end = time.time()

    print(f"Completed in {end - start:.3f} seconds.")
    print("C[0][0] =", C[0][0])

if __name__ == "__main__":
    main()

File: examples/matmul_example.c
#include <stdio.h>
#include <stdlib.h>

#define N 32 

void matmul(float A[N][N], float B[N][N], float C[N][N]) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            float sum = 0.0f;
            for (int k = 0; k < N; k++) {
                sum += A[i][k] * B[k][j];
            }
            C[i][j] = sum;
        }
    }
}

int main() {
    static float A[N][N], B[N][N], C[N][N];

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            A[i][j] = 1.0f;
            B[i][j] = 1.0f;
            C[i][j] = 0.0f;
        }
    }

    printf("Starting matrix multiplication of size %dx%d...\n", N, N);
    matmul(A, B, C);
    printf("Matrix multiplication execution completed on CPU.\n");

    return 0;
}
